src/potentials.cpp

```diff
@@ -504,92 +504,171 @@ void FunctorPotential::from_json(const json &j) {
 
 FunctorPotential::FunctorPotential(const std::string &name) : PairPotentialBase(name) {}
 
-//---------------- TabulatedPotential ---------------------
-
-void TabulatedPotential::from_json(const json &j) {
-    FunctorPotential::from_json(j);
-
-    // if user specifies an anisotropic potential, make sure to bail out
-    if (not isotropic)
-        throw std::runtime_error("cannot spline anisotropic potentials");
-
-    spline.setTolerance(j.value("utol", 1e-5), j.value("ftol", 1e-2));
-    double u_at_rmin = j.value("u_at_rmin", 20);
-    double u_at_rmax = j.value("u_at_rmax", 1e-6);
-    hardsphere = j.value("hardsphere", false);
-
-    // build matrix of spline data, each element corresponding
-    // to a pair of atom types
-    for (size_t i = 0; i < atoms.size(); ++i) {
-        for (size_t k = 0; k <= i; ++k) {
-            if (atoms[i].implicit == false and atoms[k].implicit == false) {
-                Particle a = atoms[i];
-                Particle b = atoms[k];
-                double rmin2 = .5 * (atoms[i].sigma + atoms[k].sigma);
-                rmin2 = rmin2 * rmin2;
-                double rmax2 = rmin2 * 100;
-                auto it = j.find("cutoff_g2g");
-                if (j.count("rmax") == 1) {
-                    rmax2 = std::pow(j.at("rmax").get<double>(), 2);
-                } else if (it != j.end()) {
-                    if (it->is_number())
-                        rmax2 = std::pow(it->get<double>(), 2);
-                    else if (it->is_object())
-                        rmax2 = std::pow(it->at("default").get<double>(), 2);
+/**
+ * For each pair of atom types a file is created containing
+ * the exact and splined pair potential as a function of distance
+ */
+void TabulatedPotential::save_potentials() {
+    for (size_t i = 0; i < Faunus::atoms.size(); ++i) { // loop over atom types
+        for (size_t j = 0; j <= i; ++j) {               // and build matrix of spline data (knots) for each pair
+            if (atoms[i].implicit || atoms[j].implicit) {
+                continue;
+            }
+            Particle particle1 = Faunus::atoms.at(i);
+            Particle particle2 = Faunus::atoms.at(j);
+            auto filename = fmt::format("{}-{}_tabulated.dat", Faunus::atoms[i].name, Faunus::atoms[j].name);
+            if (std::ofstream stream(filename); stream) {
+                stream << "# r u_splined/kT u_exact/kT\n";
+                double rmax = sqrt(matrix_of_knots(i, j).rmax2);
+                for (double r = dr; r < rmax; r += dr) {
+                    stream << fmt::format("{:.6E} {:.6E} {:.6E}\n",
+                                          r, operator()(particle1, particle2, r *r, {r, 0, 0}),
+                                          TabulatedPotential::operator()(particle1, particle2, r *r, {r, 0, 0}));
```

Comment:
Yes, it's annoyingly the way clang-format wants it. Not sure there's a point in changing it as it will attack it again later. Best thing would be to fix it in the clang-format config file, but it doesn't look inviting.